File: database/src/main/java/org/difft/app/database/models/GroupModel.java
package org.difft.app.database.models;

import com.tencent.wcdb.WCDBField;
import com.tencent.wcdb.WCDBIndex;
import com.tencent.wcdb.WCDBTableCoding;

import java.util.Objects;

@WCDBTableCoding
public class GroupModel {
    @WCDBField(isPrimary = true, isAutoIncrement = true)
    public long databaseId;

    @WCDBIndex(isUnique = true)
    @WCDBField
    public String gid;

    @WCDBField
    public String name;

    @WCDBField
    public Integer messageExpiry;

    @WCDBField
    public String avatar;

    @WCDBField
    public Integer status;

    @WCDBField
    public Integer invitationRule;

    @WCDBField
    public Integer version;

    @WCDBField
    public String remindCycle;

    @WCDBField
    public Boolean anyoneRemove;

    @WCDBField
    public Boolean rejoin;

    @WCDBField
    public Integer publishRule;

    @WCDBField
    public Boolean linkInviteSwitch = false;

    @WCDBField
    public Boolean privateChat = false;

    @WCDBField
    public boolean criticalAlert = false;

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        GroupModel that = (GroupModel) o;
        return Objects.equals(gid, that.gid) && Objects.equals(name, that.name) && Objects.equals(messageExpiry, that.messageExpiry) && Objects.equals(avatar, that.avatar) && Objects.equals(status, that.status) && Objects.equals(invitationRule, that.invitationRule) && Objects.equals(version, that.version) && Objects.equals(remindCycle, that.remindCycle) && Objects.equals(anyoneRemove, that.anyoneRemove) && Objects.equals(rejoin, that.rejoin) && Objects.equals(publishRule, that.publishRule) && Objects.equals(linkInviteSwitch, that.linkInviteSwitch) && Objects.equals(privateChat, that.privateChat) && criticalAlert == that.criticalAlert;
    }

    @Override
    public int hashCode() {
        return Objects.hash(gid, name, messageExpiry, avatar, status, invitationRule, version, remindCycle, anyoneRemove, rejoin, publishRule, linkInviteSwitch, privateChat, criticalAlert);
    }
}

File: database/src/main/java/org/difft/app/database/models/GroupModel.kt
package org.difft.app.database.models

import com.tencent.wcdb.WCDBField
import com.tencent.wcdb.WCDBIndex
import com.tencent.wcdb.WCDBTableCoding
import java.util.Objects

@WCDBTableCoding
class GroupModel {
    // Int (not Long): WCDB-KSP generates `== 0` autoincrement guard which does not
    // compile for Long primary keys. Column type stays ColumnType.Integer — no schema
    // change. Local rowid never exceeds 2^31 and is never externally serialized. #901
    @WCDBField(isPrimary = true, isAutoIncrement = true)
    var databaseId: Int = 0

    @WCDBIndex(isUnique = true)
    @WCDBField
    var gid: String? = null

    @WCDBField
    var name: String? = null

    @WCDBField
    var messageExpiry: Int? = null

    @WCDBField
    var avatar: String? = null

    @WCDBField
    var status: Int? = null

    @WCDBField
    var invitationRule: Int? = null

    @WCDBField
    var version: Int? = null

    @WCDBField
    var remindCycle: String? = null

    @WCDBField
    var anyoneRemove: Boolean? = null

    @WCDBField
    var rejoin: Boolean? = null

    @WCDBField
    var publishRule: Int? = null

    @WCDBField
    var linkInviteSwitch: Boolean? = false

    @WCDBField
    var privateChat: Boolean? = false

    @WCDBField
    var criticalAlert: Boolean = false

    @WCDBField
    var groupCryptoMode: Int? = null

    @WCDBField
    var encryptedName: String? = null

    @WCDBField
    var encryptedAvatar: String? = null

    @WCDBField
    var groupCryptoKeyVersion: Int? = null

    // exclude databaseId: rowid must not affect content equality (faithful to the
    // original Java equals/hashCode, which already excluded it), #901
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        other as GroupModel
        return gid == other.gid &&
                name == other.name &&
                messageExpiry == other.messageExpiry &&
                avatar == other.avatar &&
                status == other.status &&
                invitationRule == other.invitationRule &&
                version == other.version &&
                remindCycle == other.remindCycle &&
                anyoneRemove == other.anyoneRemove &&
                rejoin == other.rejoin &&
                publishRule == other.publishRule &&
                linkInviteSwitch == other.linkInviteSwitch &&
                privateChat == other.privateChat &&
                criticalAlert == other.criticalAlert &&
                groupCryptoMode == other.groupCryptoMode &&
                encryptedName == other.encryptedName &&
                encryptedAvatar == other.encryptedAvatar &&
                groupCryptoKeyVersion == other.groupCryptoKeyVersion
    }

    // exclude databaseId: rowid must not affect content equality, #901
    override fun hashCode(): Int = Objects.hash(
        gid,
        name,
        messageExpiry,
        avatar,
        status,
        invitationRule,
        version,
        remindCycle,
        anyoneRemove,
        rejoin,
        publishRule,
        linkInviteSwitch,
        privateChat,
        criticalAlert,
        groupCryptoMode,
        encryptedName,
        encryptedAvatar,
        groupCryptoKeyVersion
    )
}
